File: 백준/Silver/15656. N과 M （7）/N과 M （7）.py
n,m = map(int, input().split())
arr = list(map(int,input().split()))
arr.sort()

ans = [0] * m
v = [0] * n

def dfs(x):
    if x == m:
        print(*ans)
        return
    for i in range(n):
        ans[x] = arr[i]
        dfs(x+1)
        # ans[x] = 0


dfs(0)

File: 백준/Silver/15656. N과 M （7）/N과 M （7）.java

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.Scanner;

public class Main {
  static int n, m;
  static int arr[];
  static int ans[];
  static boolean visited[];

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    n = sc.nextInt();
    m = sc.nextInt();
    arr = new int[n];
    ans = new int[m];
    visited = new boolean[n];

    for (int i = 0; i < n; i++) {
      arr[i] = sc.nextInt();
    }
    Arrays.sort(arr);
    recur(0);
    System.out.println(sb);


  }
  static StringBuilder sb = new StringBuilder();
  private static void recur(int dept) {
    if(dept == m){
      for (int i = 0; i < m; i++) {
        sb.append(ans[i] +" ");
      }
      sb.append("\n");
      return;
    }
    for (int i = 0; i < arr.length; i++) {
      ans[dept] = arr[i];
      recur(dept + 1);
    }
  }
}
